vortex-duckdb/src/convert/types/from.rs

```diff
@@ -0,0 +1,67 @@
+use std::sync::Arc;
+
+use duckdb::core::{LogicalTypeHandle, LogicalTypeId};
+use vortex_dtype::Nullability::Nullable;
+use vortex_dtype::{DType, Nullability, PType, StructDType};
+
+pub trait FromDuckDBType<A> {
+    // Nullable is inferred from the `NotNullConstraint`.
+    fn from_duckdb(array: A, nullable: Nullability) -> Self;
+}
+
+impl FromDuckDBType<LogicalTypeHandle> for DType {
+    // Converts a DuckDB logical type handle to a `DType` based on the logical type ID.
+    fn from_duckdb(type_: LogicalTypeHandle, nullable: Nullability) -> Self {
+        match type_.id() {
+            LogicalTypeId::Boolean => DType::Bool(nullable),
+            LogicalTypeId::Tinyint => DType::Primitive(PType::I8, nullable),
+            LogicalTypeId::Smallint => DType::Primitive(PType::I16, nullable),
+            LogicalTypeId::Integer => DType::Primitive(PType::I32, nullable),
+            LogicalTypeId::Bigint => DType::Primitive(PType::I64, nullable),
+            LogicalTypeId::UTinyint => DType::Primitive(PType::U8, nullable),
+            LogicalTypeId::USmallint => DType::Primitive(PType::U16, nullable),
+            LogicalTypeId::UInteger => DType::Primitive(PType::U32, nullable),
+            LogicalTypeId::UBigint => DType::Primitive(PType::U64, nullable),
+            LogicalTypeId::Float => DType::Primitive(PType::F32, nullable),
+            LogicalTypeId::Double => DType::Primitive(PType::F64, nullable),
+            LogicalTypeId::Varchar => DType::Utf8(nullable),
+            LogicalTypeId::Blob => DType::Binary(nullable),
+            LogicalTypeId::Struct => DType::Struct(Arc::new(from_duckdb_struct(type_)), nullable),
+            LogicalTypeId::List => DType::List(Arc::new(from_duckdb_list(type_)), nullable),
+            LogicalTypeId::Timestamp
+            | LogicalTypeId::Date
+            | LogicalTypeId::Time
+            | LogicalTypeId::Interval
+            // Hugeint is a i128
+            | LogicalTypeId::Hugeint
+            | LogicalTypeId::Decimal
+            | LogicalTypeId::TimestampS
+            | LogicalTypeId::TimestampMs
+            | LogicalTypeId::TimestampNs
+            | LogicalTypeId::Enum
+            | LogicalTypeId::Map
+            | LogicalTypeId::Uuid
+            | LogicalTypeId::Union
+            | LogicalTypeId::TimestampTZ => todo!(),
+        }
+    }
+}
+
+fn from_duckdb_list(list: LogicalTypeHandle) -> DType {
```

Comment:
Same here, could consider making this more rustic: 
`impl From<LogicalTypeHandle>`..

Reply:
I am not sure we can, since we I need a nullability above and I would need a dummy struct to control the from impl here